0-module-and-module-source.emu

```diff
@@ -537,27 +577,25 @@ location: https://tc39.es/proposal-compartments/
         </ul>
 
         <emu-clause id="sec-module">
-            <h1>Module ( _moduleSource_, _referral_, _importHook_ [, _importMeta_ ] )</h1>
+            <h1>Module ( _moduleSource_, _handler_ )</h1>
             <p>When the `Module` function is called with the arguments _moduleSource_, _referral_, _importHook_ and _importMeta_, the following steps are taken:</p>
             <emu-alg>
                 1. If NewTarget is *undefined*, throw a *TypeError* exception.
-                1. If Type(_moduleSource_) is not Object, throw a *TypeError* exception.
-                1. If IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
-                1. If _importMeta_ is *undefined*, then
-                  1. Set _importMeta_ to *null*.
-                1. Else, if Type(_importMeta_) is not Object, then
-                  1. Throw a *TypeError* exception.
                 1. Perform ? RequireInternalSlot(_moduleSource_, [[ModuleSource]]).
-                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[ImportHook]] »).
                 1. Let _moduleRecord_ to ! CreateModuleRecord(_moduleSource_.[[ModuleSource]]).
+                1. If _handler_ is *null*, throw a *TypeError* exception.
+                1. If Type(_handler_) is Object, then
+                  1. Let _importHook_ be ? GetMethod(_handler_, *"importHook"*).
+                  1. If _importHook_ is not *undefined* and IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
+                  1. Let _importMetaHook_ be ? GetMethod(_handler_, *"importMetaHook"*).
+                  1. If _importMetaHook_ is not *undefined* and IsCallable(_importMetaHook_) is *false*, throw a *TypeError* exception.
+                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[HandlerValue]], [[ImportHook]], [[ImportMetaHook]] »).
+                1. Set _O_.[[ModuleSourceInstance]] to _moduleSource_.
+                1. Set _O_.[[HandlerValue]] be _handler_.
+                1. Set _O_.[[ImportHook]] be _importHook_.
+                1. Set _O_.[[ImportMetaHook]] be _importMetaHook_.
```

Comment:
1. I don't think we want to allow non-object handlers
2. `_importHook_` and `_importMetaHook_` are not initialized if `_handler_` is not an object

```suggestion
                1. If Type(_handler_) is Object, then
                  1. Let _importHook_ be ? GetMethod(_handler_, *"importHook"*).
                  1. If _importHook_ is not *undefined* and IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
                  1. Let _importMetaHook_ be ? GetMethod(_handler_, *"importMetaHook"*).
                  1. If _importMetaHook_ is not *undefined* and IsCallable(_importMetaHook_) is *false*, throw a *TypeError* exception.
                1. Else if _handler_ is *undefined*, then
                  1. Let _importHook_ be *undefined*.
                  1. Let _importMetaHook_ be *undefined*.
                1. Else,
                  1. Throw a *TypeError* exception.
                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[HandlerValue]], [[ImportHook]], [[ImportMetaHook]] »).
                1. Set _O_.[[ModuleSourceInstance]] to _moduleSource_.
                1. Set _O_.[[HandlerValue]] be _handler_.
                1. Set _O_.[[ImportHook]] be _importHook_.
                1. Set _O_.[[ImportMetaHook]] be _importMetaHook_.
```